lib/solr.ts

```diff
@@ -781,14 +611,12 @@ export class Client {
    */
   doQuery(
     handler: string,
-    query: Collection | Query | Record<string, any> | string | CallbackFn,
-    callback?: CallbackFn
+    query: Collection | Query | Record<string, any> | string,
+    callback: CallbackFn | undefined
```

Comment:
After this change, and until we implement returning Promise for cases when callback is not set, this invites user to pass incorrect params, with query and without callback. Previously callback was only optional because user potentially could pass callback instead of a query.
Also query should be an optional parameter now, because previously it could be replaced with (mandatory) callback.

I have mixed feelings about introducing such hugely breaking changes, even if polymorphism is bad both for performance and for code readability reasons, but I can imagine large codebases relying on past behaviour. I guess it does make sense, though.

Reply:
I see your point. But that's why we have semantic versioning, right? This PR and others like it would trigger a major release. And we could write a nice little "migrating to 1.x" document.

I really want to get rid of callbacks completely, and use proper async functions. This is Typescript, after all.

Reply:
Fair point. If we are dropping callbacks altogether (which work fine with TypeScript, though), then we don't need polymorphism. Let's go for it.